# 01 — Data curation: 2025 exploration and the master table

**Part 1 (explore the 2025 Domain data).** A close look at the course dataset: schema, duplicates, types, dates,
rent outliers, property types, room counts, coordinates, missing values and the feature tags. Its findings are the
evidence behind the cleaning rules (e.g. the $808,500 "rent", car spaces at $40, the empty land-area column).
Part 1 inspects; it no longer saves a table of its own.

**Part 2 (build the master table for every year).** Applies one set of rules, written once in
`scripts/listings.py`, to the 2025 Domain data and the 2026 rent.com.au scrape, and saves
`data/curated/listings.parquet`. Every later notebook reads that table.

Inputs:
- `data/raw/domain/Data/vic_rentals_all.csv` (Domain.com.au, Sept 2025, course dataset)
- `data/raw/rentcomau/2026-09/vic_rentals_rentcomau.csv` (rent.com.au, Sept 2026, from `02_scraping`; optional)

Output:
- `data/curated/listings.parquet`: one row per listing, both years stacked, 62 columns

The listing data is restricted course data (and scraped data): it is never committed to Git, and only summaries
are shown in Part 2.

# Part 1 · Exploring the 2025 Domain data

## Imports 


In [1]:
from pathlib import Path
import re

import numpy as np

import pandas as pd

## Define Paths

In [2]:
RAW_PATH = Path("../data/raw/domain/Data/vic_rentals_all.csv")   # course dataset, Sept 2025

assert RAW_PATH.exists(), f"Raw data not found: {RAW_PATH}"

## Load raw data 

In [3]:
raw = pd.read_csv(RAW_PATH)

print(f"Rows: {raw.shape[0]:,}")
print(f"Columns: {raw.shape[1]}")

Rows: 12,717
Columns: 30


## Inpsect 

In [4]:
df = raw.copy()

pd.DataFrame({
    "dtype": df.dtypes,
    "missing": df.isna().sum(),
    "missing_pct": (df.isna().mean() * 100).round(2),
    "unique": df.nunique(dropna=True)
})

,dtype,missing,missing_pct,unique
listing_id,int64,0,0.00,12717
suburb,str,0,0.00,656
postcode,int64,0,0.00,426
weekly_rent,float64,279,2.19,408
bond,float64,782,6.15,983
available_date,str,136,1.07,734
date_listed,str,4,0.03,655
days_listed,float64,4,0.03,661
bedrooms,float64,125,0.98,11
bathrooms,float64,51,0.40,9


### Initial observations

- `listing_id` is complete and unique and can be used as the listing-level identifier.
- Rental and core property variables have relatively low missingness.
- `carspaces` and `structured_features` have moderate missingness.
- `property_id` has substantial missingness and is unsuitable as the primary identifier.
- `land_area` is almost entirely missing and is unlikely to be useful downstream.
- Date columns require conversion from strings to datetime.
- `postcode` should be treated as a categorical/location identifier rather than a numeric measurement.
- Latitude and longitude are almost complete, making spatial assignment to SA2 feasible.

## Validate required columns

In [5]:
required_columns = {
    "listing_id",
    "suburb",
    "postcode",
    "weekly_rent",
    "bond",
    "available_date",
    "date_listed",
    "days_listed",
    "bedrooms",
    "bathrooms",
    "carspaces",
    "property_type",
    "address",
    "lat",
    "lon",
    "scraped_date",
    "structured_features"
}

missing_columns = required_columns - set(df.columns)

assert not missing_columns, f"Missing required columns: {missing_columns}"

## Check duplicates

In [6]:
print("Exact duplicate rows:", df.duplicated().sum())
print("Duplicate listing IDs:", df["listing_id"].duplicated().sum())
print("Unique listing IDs:", df["listing_id"].nunique())


Exact duplicate rows: 0
Duplicate listing IDs: 0
Unique listing IDs: 12717


## Standardise text fields

In [7]:
text_columns = [
    "suburb",
    "address",
    "property_type",
    "primary_type",
    "secondary_type",
    "agency",
    "agent_names"
]

for col in text_columns:
    if col in df.columns:
        df[col] = df[col].astype("string").str.strip()

## Standardise postcode

In [8]:
df["postcode"] = (
    pd.to_numeric(df["postcode"], errors="coerce")
    .astype("Int64")
    .astype("string")
    .str.zfill(4)
)

## Clean numeric columns

In [9]:
numeric_columns = [
    "weekly_rent",
    "bond",
    "days_listed",
    "bedrooms",
    "bathrooms",
    "carspaces",
    "photo_count",
    "video_count",
    "floorplans_count",
    "lat",
    "lon"
]

for col in numeric_columns:
    if col in df.columns:
        df[col] = pd.to_numeric(df[col], errors="coerce")

## Convert integer-like fields

In [10]:
integer_columns = [
    "bedrooms",
    "bathrooms",
    "carspaces",
    "photo_count",
    "video_count",
    "floorplans_count",
    "days_listed"
]

for col in integer_columns:
    if col in df.columns:
        df[col] = df[col].round().astype("Int64")

## Parse dates

In [11]:
df["date_listed"] = pd.to_datetime(
    df["date_listed"],
    errors="coerce"
)

df["scraped_date"] = pd.to_datetime(
    df["scraped_date"],
    errors="coerce"
)

df["available_date"] = pd.to_datetime(
    df["available_date"],
    errors="coerce"
)

df[
    ["date_listed", "available_date", "scraped_date"]
].describe()

,date_listed,available_date,scraped_date
count,12713,12581,12717
mean,2025-06-21 22:43:18.977424,2025-07-10 08:37:41.696208,2025-09-09 16:41:58.622867
min,2008-01-15 00:00:00,1970-01-01 00:00:00,2025-09-09 16:13:04
25%,2025-07-30 00:00:00,2025-08-13 00:00:00,2025-09-09 16:26:02
50%,2025-08-21 00:00:00,2025-09-03 00:00:00,2025-09-09 16:39:08
75%,2025-09-02 00:00:00,2025-09-19 00:00:00,2025-09-09 16:53:58
max,2025-09-09 00:00:00,2026-10-03 00:00:00,2025-09-09 18:07:20


## Recalculate days listed 

In [12]:
df["days_listed_calc"] = (
    df["scraped_date"].dt.normalize()
    - df["date_listed"].dt.normalize()
).dt.days

comparison = (
    df["days_listed"].astype("Float64")
    - df["days_listed_calc"].astype("Float64")
)

comparison.value_counts(dropna=False).head()

 0.0     10813
-1.0      1900
 <NA>        4
Name: count, dtype: Int64

### Days-listed validation

The supplied `days_listed` field was compared with the calendar-day difference between `scraped_date` and `date_listed`.

- 10,813 observations matched exactly.
- 1,900 observations differed by exactly one day.
- 4 observations could not be compared due to missing listing dates.

The one-day discrepancies are consistent with the loss of time-of-day information in `date_listed`. The supplied `days_listed` field is therefore retained.

## Flag long/impossible dates

In [13]:
df["flag_future_listed"] = (
    df["date_listed"] > df["scraped_date"]
)

df["flag_long_listing"] = df["days_listed"] > 365

df["flag_invalid_available_date"] = (
    df["available_date"].dt.year < 2020
)

## Inspect rental prices

In [14]:
df["weekly_rent"].describe(
    percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99]
)

count     12438.000000
mean        764.713378
std        9428.719738
min           0.000000
1%          221.850000
5%          360.000000
25%         485.000000
50%         560.000000
75%         690.000000
95%        1100.000000
99%        1750.000000
max      808500.000000
Name: weekly_rent, dtype: float64

The weekly rent distribution is strongly right-skewed. While the median weekly rent is $560 and 99% of listings are below $1,750, the maximum observed value is $808,500. These extreme values substantially inflate the mean and standard deviation and are unlikely to represent genuine weekly residential rents. Suspicious rent observations should therefore be flagged and inspected before downstream analysis, while robust statistics such as the median are preferred for summarising rental prices.


In [15]:
df.nlargest(
    30,
    "weekly_rent"
)[
    [
        "listing_id",
        "weekly_rent",
        "property_type",
        "secondary_type",
        "bedrooms",
        "suburb",
        "address",
        "url"
    ]
]

,listing_id,weekly_rent,property_type,secondary_type,bedrooms,suburb,address,url
2586,17697966,808500.0,House,House,4,BERWICK,83 Golf Links Road,https://www.domain.com.au/83-golf-links-road-b...
3641,17750082,595000.0,House,House,3,ROSEBUD,4 Wilfred Street,https://www.domain.com.au/4-wilfred-street-ros...
1929,17747509,315000.0,Apartment / Unit / Flat,Apartment / Unit / Flat,1,MORWELL,1/28 Elgin Street,https://www.domain.com.au/1-28-elgin-street-mo...
10356,17712191,10000.0,House,House,6,TOORAK,<NA>,https://www.domain.com.au/toorak-vic-3142-1771...
12294,17661725,5866.0,House,House,5,CAMBERWELL,54 Fairmont Avenue,https://www.domain.com.au/54-fairmont-avenue-c...
10358,17722803,5750.0,House,House,4,TOORAK,2 Lisbuoy Court,https://www.domain.com.au/2-lisbuoy-court-toor...
12705,9287841,5000.0,Apartment / Unit / Flat,Apartment / Unit / Flat,5,BRIGHTON,<NA>,https://www.domain.com.au/brighton-vic-3186-92...
8776,17744678,4500.0,House,House,5,HAWTHORN,94 Illawarra Road,https://www.domain.com.au/94-illawarra-road-ha...
10768,17694334,4500.0,Apartment / Unit / Flat,Apartment / Unit / Flat,3,SOUTH YARRA,1 Fairlie Court,https://www.domain.com.au/1-fairlie-court-sout...
1236,17460975,4000.0,Apartment / Unit / Flat,Apartment / Unit / Flat,4,SOUTH MELBOURNE,901/161 Eastern Road,https://www.domain.com.au/901-161-eastern-road...


Inspection of the upper tail shows that most high weekly rents are associated with plausible premium properties in suburbs such as Toorak, Brighton, South Yarra, Docklands, and Port Melbourne. Values in the range of approximately $3,000–$10,000 per week may therefore represent genuine luxury rental listings and should not be removed solely because they are large.

However, three observations — $315,000, $595,000, and $808,500 per week — are orders of magnitude larger than the rest of the distribution and are inconsistent with plausible weekly residential rents. These values likely reflect data-entry or scraping errors, such as sale prices being captured as rental prices. They should be flagged for exclusion or correction in downstream rent-based analysis rather than applying a blanket upper threshold to all expensive listings.


In [16]:
df.nsmallest(
    30,
    "weekly_rent"
)[
    [
        "listing_id",
        "weekly_rent",
        "property_type",
        "secondary_type",
        "bedrooms",
        "suburb",
        "address",
        "url"
    ]
]

,listing_id,weekly_rent,property_type,secondary_type,bedrooms,suburb,address,url
3111,17159699,0.0,House,House,3,CLYDE,24 Viola Circuit,https://www.domain.com.au/24-viola-circuit-cly...
4353,10344261,0.0,House,House,2,NORTHCOTE,40 Charles Street,https://www.domain.com.au/40-charles-street-no...
5087,16697431,0.0,House,House,2,QUARRY HILL,11 Pyke Street,https://www.domain.com.au/11-pyke-street-quarr...
5091,17175722,0.0,House,House,2,QUARRY HILL,158 Mitchell Street,https://www.domain.com.au/158-mitchell-street-...
5092,17442602,0.0,House,House,4,QUARRY HILL,30 Echuca Street,https://www.domain.com.au/30-echuca-street-qua...
5887,16706257,0.0,House,House,4,NORTH BENDIGO,16 Michelsen Street,https://www.domain.com.au/16-michelsen-street-...
5892,17416575,0.0,Townhouse,Townhouse,3,NORTH BENDIGO,1/28 Bayne Street,https://www.domain.com.au/1-28-bayne-street-no...
7012,15367349,0.0,House,House,4,MERNDA,21 William Street,https://www.domain.com.au/21-william-street-me...
7014,16248581,0.0,House,House,3,MERNDA,8 Muir Way,https://www.domain.com.au/8-muir-way-mernda-vi...
8305,17316329,0.0,House,House,2,BENDIGO,221 Barnard Street,https://www.domain.com.au/221-barnard-street-b...


Inspection of the lower tail shows that very low weekly rents have mixed interpretations. Several observations with rents between approximately $35 and $50 correspond to car spaces or addresses explicitly referring to parking, so these values may be legitimate despite appearing unusually low.

In contrast, values such as $0–$40 attached to houses, townhouses, or apartments are implausible as normal residential weekly rents and are likely to represent missing, incorrectly parsed, or otherwise invalid rental prices. This indicates that low-rent validation should consider both the rental value and the listing type rather than applying a single numerical cutoff across all listings.

## Inspect non-standard property listings


In [17]:
df["secondary_type"].value_counts(dropna=False)

secondary_type
House                            6587
Apartment / Unit / Flat          4516
Townhouse                        1276
Studio                            209
New Apartments / Off the Plan      45
Villa                              28
Acreage / Semi-Rural               13
Semi-Detached                      10
New House & Land                    8
Car Space                           7
<NA>                                4
Duplex                              3
Block of Units                      3
Terrace                             3
Farm                                2
retirement                          2
Vacant land                         1
Name: count, dtype: int64[pyarrow]

In [18]:
df["primary_type"].value_counts(dropna=False)

primary_type
House               6598
Apartment           4728
Townhouse/Villa     1304
New Developments      45
Any                   37
<NA>                   4
Land                   1
Name: count, dtype: int64[pyarrow]

In [19]:
standard_residential_types = {
    "House",
    "Apartment / Unit / Flat",
    "Townhouse",
    "Studio",
    "Villa",
    "Semi-Detached",
    "Duplex",
    "Block of Units",
    "Terrace"
}

df["is_standard_residential"] = (
    df["secondary_type"].isin(standard_residential_types)
)

In [20]:
df["is_standard_residential"].value_counts()

is_standard_residential
True     12635
False       82
Name: count, dtype: int64

## Flag suspicious rents

In [21]:
df["flag_rent_suspicious"] = (
    df["is_standard_residential"]
    & (
        (df["weekly_rent"] < 100)
        | (df["weekly_rent"] > 10_000)
    )
)

## Inpsect bedrooms, bathrooms, and carspaces

In [22]:
for col in ["bedrooms", "bathrooms", "carspaces"]:
    print(f"\n{col}")
    print(
        df[col]
        .value_counts(dropna=False)
        .sort_index()
    )


bedrooms
bedrooms
1       1834
2       3416
3       4054
4       2973
5        261
6         28
7         10
8          8
9          6
11         1
50         1
<NA>     125
Name: count, dtype: Int64

bathrooms
bathrooms
1       5991
2       6026
3        575
4         55
5          9
6          3
7          1
9          5
12         1
<NA>      51
Name: count, dtype: Int64

carspaces
carspaces
1       5099
2       5023
3        369
4        294
5         52
6         36
7          6
8         10
9          4
10         9
11         2
12         3
14         2
15         1
19         1
20         2
22         1
<NA>    1803
Name: count, dtype: Int64


In [23]:
df.loc[
    (df["bedrooms"] >= 10)
    | (df["bathrooms"] >= 8)
    | (df["carspaces"] >= 10),
    [
        "listing_id",
        "weekly_rent",
        "bedrooms",
        "bathrooms",
        "carspaces",
        "property_type",
        "secondary_type",
        "suburb",
        "address",
        "url"
    ]
].sort_values(
    ["bedrooms", "bathrooms", "carspaces"],
    ascending=False
)

,listing_id,weekly_rent,bedrooms,bathrooms,carspaces,property_type,secondary_type,suburb,address,url
5362,15800260,1050.0,50,1,1,Apartment / Unit / Flat,Apartment / Unit / Flat,SOUTHBANK,1 Freshwater Place,https://www.domain.com.au/1-freshwater-place-s...
2133,17748261,1800.0,11,6,5,House,House,WOODSTOCK,1300 Donnybrook Road,https://www.domain.com.au/1300-donnybrook-road...
6949,17027842,3300.0,9,9,8,House,House,NEWTOWN,<NA>,https://www.domain.com.au/newtown-vic-3220-170...
2011,17737763,300.0,9,9,3,House,House,WAURN PONDS,27 Danaher Avenue,https://www.domain.com.au/27-danaher-avenue-wa...
7423,17735877,375.0,9,9,<NA>,House,House,WERRIBEE,78 Market Rd,https://www.domain.com.au/78-market-rd-werribe...
10347,17288084,NaN,6,5,10,House,House,TOORAK,<NA>,https://www.domain.com.au/toorak-vic-3142-1728...
11259,17735862,2000.0,5,4,15,House,House,WARRANDYTE,287-301 Jumping Creek Rd,https://www.domain.com.au/287-301-jumping-cree...
10464,17739856,895.0,5,2,12,House,House,NAR NAR GOON,5 Latta Road,https://www.domain.com.au/5-latta-road-nar-nar...
5599,16991827,650.0,5,2,10,House,House,CRAIGIEBURN,6 Plumpton Avenue,https://www.domain.com.au/6-plumpton-avenue-cr...
10491,17510840,1150.0,4,2,11,House,House,RYE,82 CREEDMORE DRIVE,https://www.domain.com.au/82-creedmore-drive-r...


## Flag suspicious property attributes

In [24]:
df["flag_bedrooms_suspicious"] = (
    df["bedrooms"] > 20
)

df["flag_bathrooms_suspicious"] = (
    (df["bathrooms"] > 10)
    | (
        (df["bathrooms"] >= 8)
        & (df["bedrooms"] <= 2)
    )
)

df["flag_carspaces_suspicious"] = (
    (df["carspaces"] >= 10)
    & df["secondary_type"].isin([
        "Apartment / Unit / Flat",
        "Studio"
    ])
)

df["flag_property_features_suspicious"] = (
    df["flag_bedrooms_suspicious"]
    | df["flag_bathrooms_suspicious"]
    | df["flag_carspaces_suspicious"]
)

property_flags = [
    "flag_bedrooms_suspicious",
    "flag_bathrooms_suspicious",
    "flag_carspaces_suspicious",
    "flag_property_features_suspicious"
]

for col in property_flags:
    print(f"\n{col}")
    print(df[col].value_counts(dropna=False))


flag_bedrooms_suspicious
flag_bedrooms_suspicious
False    12591
<NA>       125
True         1
Name: count, dtype: Int64

flag_bathrooms_suspicious
flag_bathrooms_suspicious
False    12663
<NA>        51
True         3
Name: count, dtype: Int64

flag_carspaces_suspicious
flag_carspaces_suspicious
False    11526
<NA>      1182
True         9
Name: count, dtype: Int64

flag_property_features_suspicious
flag_property_features_suspicious
False    11456
<NA>      1248
True        13
Name: count, dtype: Int64


In [25]:
df.loc[
    df["flag_property_features_suspicious"],
    [
        "listing_id", "bedrooms", "bathrooms", "carspaces",
        "secondary_type", "suburb", "address"
    ] + property_flags
]

,listing_id,bedrooms,bathrooms,carspaces,secondary_type,suburb,address,flag_bedrooms_suspicious,flag_bathrooms_suspicious,flag_carspaces_suspicious,flag_property_features_suspicious
1279,17030693,1,9,<NA>,House,BELL POST HILL,51 Braund Avenue,False,True,False,True
4299,8431375,1,1,20,Apartment / Unit / Flat,BURWOOD,390 Burwood Highway,False,False,True,True
4300,8431378,1,1,12,Apartment / Unit / Flat,BURWOOD,224 Burwood Highway,False,False,True,True
4301,8431379,1,1,11,Apartment / Unit / Flat,BURWOOD,308 Burwood Highway,False,False,True,True
5362,15800260,50,1,1,Apartment / Unit / Flat,SOUTHBANK,1 Freshwater Place,True,False,False,True
6353,10621148,1,1,19,Studio,ST ALBANS,1A & 1B/101-103 Main Road West,False,False,True,True
6821,8431398,<NA>,1,10,Apartment / Unit / Flat,HAWTHORN EAST,2 Eastern Place,<NA>,False,True,True
6909,17753508,1,12,8,Studio,GEELONG WEST,6 Aberdeen St,False,True,False,True
6948,16916036,1,9,8,House,NEWTOWN,258 Latrobe Terrace,False,True,False,True
8827,8431400,<NA>,1,14,Apartment / Unit / Flat,HAWTHORN,1 Queens Avenue,<NA>,False,True,True


The 50-bedroom apartment and one-bedroom listings with very high bathroom or parking counts warrant review. Large houses can reasonably have many bedrooms, bathrooms or car spaces, so contextual flags are preferable to blanket cutoffs. All rows are retained; flags remain unknown where missing attributes prevent a decision.

## Coordinate validation

In [26]:
df[["lat", "lon"]].describe()

,lat,lon
count,12713.000000,12713.000000
mean,-37.781359,144.971809
std,0.388812,0.569393
min,-38.829080,141.000550
25%,-37.904865,144.870390
50%,-37.823673,144.982800
75%,-37.750935,145.111970
max,-34.166810,149.756800


In [27]:
df.loc[
    df[["lat", "lon"]].isna().any(axis=1),
    ["listing_id", "address", "suburb", "lat", "lon"]
]

,listing_id,address,suburb,lat,lon
10179,17727626,<NA>,Rowville,NaN,NaN
11405,17721987,<NA>,Irymple,NaN,NaN
11605,17722092,<NA>,Coimadai,NaN,NaN
11606,17727546,<NA>,Coimadai,NaN,NaN


In [28]:
df["flag_bad_coordinates"] = (
    df["lat"].isna()
    | df["lon"].isna()
    | ~df["lat"].between(-40, -33)
    | ~df["lon"].between(140, 150)
)

print(df["flag_bad_coordinates"].value_counts(dropna=False))
df.loc[
    df["flag_bad_coordinates"],
    ["listing_id", "address", "suburb", "lat", "lon"]
]

flag_bad_coordinates
False    12713
True         4
Name: count, dtype: int64


,listing_id,address,suburb,lat,lon
10179,17727626,<NA>,Rowville,NaN,NaN
11405,17721987,<NA>,Irymple,NaN,NaN
11605,17722092,<NA>,Coimadai,NaN,NaN
11606,17727546,<NA>,Coimadai,NaN,NaN


Coordinates are nearly complete. This broad bounding-box check identifies missing or out-of-range values; it does not establish that each point lies within Victoria. Coordinates will support SA2 spatial assignment in the dedicated geospatial notebook.

## Land area

In [29]:
print(df["land_area"].describe())
print(f"Missing land area: {df['land_area'].isna().mean():.2%}")
df.loc[df["land_area"].notna(), ["listing_id", "land_area"]]

count          2
unique         2
top       189 m²
freq           1
Name: land_area, dtype: object
Missing land area: 99.98%


,listing_id,land_area
4075,17725252,189 m²
12109,17733960,530 m2


Only two listings have land area recorded, leaving approximately 99.98% missing. Coverage is insufficient for meaningful downstream analysis, so this variable is removed.

In [30]:
df = df.drop(columns=["land_area"])

## Missing-value policy

In [31]:
df["features_missing"] = df["structured_features"].isna()

df["coordinates_missing"] = (
    df[["lat", "lon"]]
    .isna()
    .any(axis=1)
)

df[["features_missing", "coordinates_missing"]].sum()

features_missing       1326
coordinates_missing       4
dtype: int64

Missing rent, bond, property counts, dates and coordinates are preserved rather than imputed or replaced with zero. Downstream analyses should filter only the variables they require. Missing feature information is recorded separately from the feature indicators.

## Virtual tour

In [32]:
df["virtual_tour"].value_counts(dropna=False)

virtual_tour
False    11907
True       806
NaN          4
Name: count, dtype: int64

In [33]:
assert df["virtual_tour"].dropna().isin([True, False]).all()
df["virtual_tour"] = df["virtual_tour"].astype("boolean")
df["virtual_tour"].value_counts(dropna=False)

virtual_tour
False    11907
True       806
<NA>         4
Name: count, dtype: Int64

Virtual tour contains only true, false and missing values. Nullable boolean preserves all three states.

## Structured features

In [34]:
df["structured_features"].dropna().head(10)

0     Built in wardrobes, Secure Parking, Bath, Heat...
1               Internal Laundry, Pets Allowed, Heating
2     Internal Laundry, Balcony / Deck, Floorboards,...
3     Air conditioning, Built in wardrobes, Balcony ...
4                           Built in wardrobes, Heating
6     Air conditioning, Heating, Bush Retreat, Ensui...
7                                      Internal Laundry
8     Broadband internet access, Air conditioning, H...
9                 Internal Laundry, Bath, Heating, Shed
10                                        Bath, Heating
Name: structured_features, dtype: str

In [35]:
print("Missing feature records:", df["structured_features"].isna().sum())
print(f"Missing features: {df['structured_features'].isna().mean():.2%}")

raw_feature_labels = (
    df["structured_features"]
    .dropna()
    .str.split(",")
    .explode()
)

raw_feature_labels.value_counts().head(30)

Missing feature records: 1326
Missing features: 10.43%


structured_features
 Dishwasher              5262
 Heating                 5103
 Built in wardrobes      4658
 Secure Parking          3616
 Internal Laundry        2366
 Balcony / Deck          2338
 Air conditioning        2054
Built in wardrobes       2032
 Floorboards             1801
Air conditioning         1742
 Bath                    1642
Internal Laundry         1627
 Intercom                1473
Gas                      1453
 Ensuite                 1425
 Fully fenced            1351
Secure Parking           1350
 Study                   1176
 Gas                      968
 Close to shops           963
 Furnished                951
 Close to schools         929
 Close to transport       901
 Remote Garage            881
 Pets Allowed             872
 Shed                     867
 Ducted Heating           852
 Garden / Courtyard       778
 Split System Heating     760
 Gym                      654
Name: count, dtype: int64

In [36]:
def normalise_feature_name(feature):
    feature = feature.strip().lower()
    feature = feature.replace("&", "and")
    feature = re.sub(r"[/\-]", " ", feature)
    feature = re.sub(r"\s+", " ", feature)
    return feature.strip()

normalised_feature_labels = raw_feature_labels.map(normalise_feature_name)
normalised_feature_labels.value_counts().head(60)

structured_features
built in wardrobes                         6690
heating                                    5710
dishwasher                                 5456
secure parking                             4966
internal laundry                           3993
air conditioning                           3796
balcony deck                               2509
gas                                        2421
floorboards                                1904
bath                                       1881
intercom                                   1629
ensuite                                    1621
fully fenced                               1433
study                                      1280
furnished                                  1128
pets allowed                               1047
close to shops                              965
close to transport                          947
split system heating                        933
close to schools                            932
remote garage       

In [37]:
feature_aliases = {
    "split system air con": "split system air conditioning",
    "split system aircon": "split system air conditioning",
    "split system airconditioning": "split system air conditioning",
    "air conditioner": "air conditioning",
    "broadband": "broadband internet access",
    "broadband internet available": "broadband internet access",
    "broadband connection": "broadband internet access",
    "deck balcony": "balcony deck"
}

features_clean = []

for features in df["structured_features"]:
    if pd.isna(features):
        features_clean.append(None)
    else:
        cleaned = []
        for feature in features.split(","):
            feature = normalise_feature_name(feature)
            feature = feature_aliases.get(feature, feature)
            if feature and feature not in cleaned:
                cleaned.append(feature)
        features_clean.append(sorted(cleaned))

df["features_clean"] = features_clean

feature_counts = df["features_clean"].explode().value_counts()
print("Raw feature labels:", raw_feature_labels.nunique())
print("Normalised feature labels:", normalised_feature_labels.nunique())
print("Clean feature labels:", len(feature_counts))
feature_counts.head(60)

Raw feature labels: 653
Normalised feature labels: 521
Clean feature labels: 513


features_clean
built in wardrobes                          6690
heating                                     5710
dishwasher                                  5456
secure parking                              4966
internal laundry                            3993
air conditioning                            3920
balcony deck                                2510
gas                                         2421
floorboards                                 1904
bath                                        1881
intercom                                    1629
ensuite                                     1621
fully fenced                                1433
study                                       1280
furnished                                   1128
pets allowed                                1047
split system air conditioning               1023
close to shops                               965
close to transport                           947
split system heating                         932
close

Case, spacing and slash/hyphen variants are normalised for labels such as built-in wardrobes, balcony/deck, secure parking and internal laundry. The alias map merges only clear wording variants. Balcony, deck and the combined balcony/deck label remain separate, as do general air conditioning and specific systems. Each cleaned feature is counted once per listing.

In [38]:
MIN_FEATURE_COUNT = 100
common_features = feature_counts[feature_counts >= MIN_FEATURE_COUNT].index
feature_columns = []

for feature in common_features:
    column = "feature_" + re.sub(r"[^a-z0-9]+", "_", feature).strip("_")
    assert column not in df.columns, f"Duplicate feature column: {column}"
    df[column] = df["features_clean"].apply(
        lambda features: int(features is not None and feature in features)
    )
    feature_columns.append(column)

print("Feature indicator columns:", len(feature_columns))
df[feature_columns].sum().sort_values(ascending=False)

Feature indicator columns: 49


feature_built_in_wardrobes               6690
feature_heating                          5710
feature_dishwasher                       5456
feature_secure_parking                   4966
feature_internal_laundry                 3993
feature_air_conditioning                 3920
feature_balcony_deck                     2510
feature_gas                              2421
feature_floorboards                      1904
feature_bath                             1881
feature_intercom                         1629
feature_ensuite                          1621
feature_fully_fenced                     1433
feature_study                            1280
feature_furnished                        1128
feature_pets_allowed                     1047
feature_split_system_air_conditioning    1023
feature_close_to_shops                    965
feature_close_to_transport                947
feature_split_system_heating              932
feature_close_to_schools                  932
feature_broadband_internet_access 

There were 653 raw feature labels. Case, punctuation and spacing normalisation reduced this to 521, and explicit aliases reduced it to 513. Only 49 features appear in at least 100 listings, avoiding hundreds of sparse indicator columns.

A zero indicator means the feature was not recorded, rather than confirmed absent. For the listings with missing feature information, all indicators are zero and `features_missing` is true. The original `structured_features` and cleaned lists are retained for traceability, including rare labels.

## Final missingness and schema check

In [39]:
# Lists are converted to tuples only for counting unique combinations.
unique_counts = df.drop(columns=["features_clean"]).nunique(dropna=True)
unique_counts["features_clean"] = df["features_clean"].apply(
    lambda features: tuple(features) if features is not None else None
).nunique(dropna=True)

missingness_summary = pd.DataFrame({
    "dtype": df.dtypes,
    "missing": df.isna().sum(),
    "missing_pct": (df.isna().mean() * 100).round(2),
    "unique": unique_counts
}).sort_values("missing_pct", ascending=False)

missingness_summary

,dtype,missing,missing_pct,unique
property_id,str,10145,79.78,2555
carspaces,Int64,1803,14.18,17
features_clean,object,1326,10.43,6866
structured_features,str,1326,10.43,7320
flag_property_features_suspicious,boolean,1248,9.81,2
...,...,...,...,...
postcode,string,0,0.00,426
property_type,string,0,0.00,16
scraped_date,datetime64[us],0,0.00,935
suburb,string,0,0.00,656


In [40]:
missingness_summary.loc[
    missingness_summary["missing"] > 0
].head(20)

,dtype,missing,missing_pct,unique
property_id,str,10145,79.78,2555
carspaces,Int64,1803,14.18,17
features_clean,object,1326,10.43,6866
structured_features,str,1326,10.43,7320
flag_property_features_suspicious,boolean,1248,9.81,2
flag_carspaces_suspicious,boolean,1182,9.29,2
bond,float64,782,6.15,983
weekly_rent,float64,279,2.19,408
agent_names,string,248,1.95,3586
available_date,datetime64[us],136,1.07,734


Property identifiers, feature records and some property counts remain incomplete. Smaller gaps in rent, bond, dates and coordinates are also preserved. Neither imputation nor dropping these listings is justified at this stage; `property_id` and `domain_page_id` remain available for later checks of repeated physical properties.

## Provenance

In [41]:
df["source"] = "Domain.com.au"
df["dataset_snapshot"] = pd.Timestamp("2025-09-09")
# Use nanosecond precision so the dtype is preserved in Parquet.
df["dataset_snapshot"] = df["dataset_snapshot"].astype("datetime64[ns]")

## Integrity checks

In [42]:
assert df["listing_id"].notna().all()
assert df["listing_id"].is_unique
assert len(df) == len(raw)

# Use tuples in a temporary copy because list values cannot be hashed.
duplicate_check = df.copy()
duplicate_check["features_clean"] = duplicate_check["features_clean"].apply(
    lambda features: tuple(features) if features is not None else None
)
assert not duplicate_check.duplicated().any()
del duplicate_check

assert df[feature_columns].isin([0, 1]).all().all()
print(f"Integrity checks passed: {len(df):,} unique listings retained.")

Integrity checks passed: 12,717 unique listings retained.


## Deterministic ordering

In [43]:
df = (
    df
    .sort_values("listing_id")
    .reset_index(drop=True)
)

df.head(30)

,listing_id,suburb,postcode,weekly_rent,bond,available_date,date_listed,days_listed,bedrooms,bathrooms,...,feature_rumpus_room,feature_ducted_cooling,feature_city_views,feature_gas_heating,feature_deck,feature_laundry,feature_separate_dining_room,feature_double_glazed_windows,source,dataset_snapshot
0,5470976,ASCOT VALE,3032,660.0,2868.0,2025-09-10,2025-07-21,50,2,1,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
1,5604062,MELTON,3337,NaN,150.0,2008-01-15,2008-01-15,6446,<NA>,<NA>,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
2,6168570,MELBOURNE,3000,310.0,1347.0,2023-07-03,2025-09-02,7,<NA>,1,...,0,0,1,0,0,0,0,1,Domain.com.au,2025-09-09
3,7117948,MOONEE PONDS,3039,500.0,2173.0,2025-09-19,2025-09-06,3,2,1,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
4,7455074,PRAHRAN,3181,NaN,340.0,2014-05-22,2012-03-02,4938,1,1,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
5,7625736,BRIGHTON,3186,1575.0,NaN,2022-06-18,2012-06-09,4839,2,2,...,0,1,0,0,0,1,0,0,Domain.com.au,2025-09-09
6,8074820,QUARRY HILL,3550,650.0,2824.0,2026-07-21,2012-12-12,4654,2,2,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
7,8253224,BENDIGO,3550,580.0,2520.0,2026-01-20,2013-03-25,4551,2,1,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09
8,8303216,ARMADALE,3143,1575.0,NaN,2022-06-18,2013-04-22,4522,2,2,...,0,1,0,0,0,1,0,0,Domain.com.au,2025-09-09
9,8430042,CLAYTON,3168,360.0,NaN,2018-07-20,2023-03-02,922,1,1,...,0,0,0,0,0,0,0,0,Domain.com.au,2025-09-09


## Part 1 summary

The 2025 dataset has one row per Domain listing, identified by `listing_id`. Nothing was deleted for unusual values;
suspicious observations were flagged. Land area was almost entirely missing. Feature tags were tidied and turned into
yes/no columns, with "no tag" kept separate from "confirmed absent". Coordinates were validated.

**How these findings carry into Part 2:**

| Part 1 finding | Rule in `scripts/listings.py` (Part 2) |
|---|---|
| Rents of $0–$40 on homes, and $315,000–$808,500 "rents" | Rent must be $100–$10,000 a week; genuine luxury rents up to $10,000 (Toorak) are kept |
| Car spaces, vacant land, farms, "Unknown" types | Every residential type is kept (including farmhouses and retirement units); car spaces, land and single rooms are not |
| Implausible room counts (e.g. a 50-bedroom apartment) | More than 10 bedrooms, or 8+ bathrooms with 2 or fewer bedrooms, is set aside |
| A few missing or out-of-range coordinates | Coordinates must fall inside Victoria |
| 653 feature labels → 49 columns learned from Domain's wording | A fixed list of 28 features matched by meaning, so 2025 and 2026 (different websites, different wording) get the same columns |
| Land area 99.98% missing | Not carried into the master table |

Part 1 no longer saves `vic_rentals.parquet`: the master table in Part 2 replaces it.

# Part 2 · The master table (every year, one set of rules)

From here the cleaning rules come from `scripts/listings.py`, so every notebook and every year is cleaned the same
way. In plain words:

1. **Translate each source.** Domain and rent.com.au name columns, write property types and word features
   differently. A small loader per source turns each into the same columns. IDs get a prefix (`domain-…`,
   `rentcomau-…`) so the two sites' numbers can't clash, and every row records its `source` and `snapshot`.
2. **Property groups.** House, Townhouse, Apartment/Unit, Studio, and "Other residential" (a listing typed
   "Other"/"Unknown" that has at least one bedroom). Car spaces, land and single rooms are not residential.
3. **Rule A.** A listing is `usable` if it is residential, its rent is $100–$10,000 a week, its coordinates are in
   Victoria, its bedroom count is known and its room counts are plausible. Nothing is deleted: listings that fail
   are kept, and `exclusion_reason` names the first check they fail. The nine review flags from Part 1
   (`flag_future_listed`, `flag_long_listing`, … `flag_bad_coordinates`) are also kept, computed the same way for
   every year: they mark listings for review, while Rule A decides what is `usable`.
4. **28 features.** Each listing's feature tags are matched by meaning to 28 yes/no columns: 1 = mentioned,
   0 = not mentioned, blank = unknown. Blank is used when a listing has no feature list at all
   (`features_missing`), and for the 10 features rent.com.au's listing form barely offers (built-in wardrobes,
   laundry, solar, gas appliances, bath, carpet, openable windows, views, separate dining, double glazing): there a
   2026 listing is "yes if mentioned, otherwise unknown", never "no". Pets, furnished and ensuite come from
   rent.com.au's own fields, so they are known for 2026 even without a feature list.
5. **Check, then save.** The table is refused if listing IDs repeat, a year is unknown, a suburb key is badly formed
   or a feature has a value other than 1, 0 or blank.

In [44]:
import sys
from pathlib import Path
sys.path.insert(0, str(Path.cwd().parent / "scripts"))   # shared project code

import pandas as pd
import config
import listings as L

## 2.1 The 2026 scrape

The merged rent.com.au file from `02_scraping` (skipped if it hasn't been created yet).

In [45]:
if config.RENTCOMAU_2026.exists():
    raw26 = pd.read_csv(config.RENTCOMAU_2026, low_memory=False)
    own = raw26["in_searched_suburb"].astype(str).str.lower().eq("true")
    print(f"rent.com.au file: {len(raw26):,} listings from {raw26['search_suburb'].nunique()} searched suburbs "
          f"({raw26['worker'].nunique()} people)")
    print(f"  found in their own suburb's search: {own.mean():.1%} "
          "(the rest came up in a neighbouring suburb's results and are kept: one copy per listing)")
    print(f"  with listing-page details (bond, features): {raw26['detail_scraped_date'].notna().mean():.1%}")
    cols = ["weekly_rent", "bedrooms", "bathrooms", "carspaces", "property_type", "lat", "lon",
            "date_listed", "agency", "bond", "structured_features", "photo_count"]
    display(raw26[cols].notna().mean().mul(100).round(1).rename("% filled").to_frame())
else:
    print("No 2026 file yet (run 02_scraping's merge first). The master table is built from 2025 only.")

No 2026 file yet (run 02_scraping's merge first). The master table is built from 2025 only.


## 2.2 Build the master table

In [46]:
master = L.build_listings()     # loads every available source, cleans, checks, saves listings.parquet

overview = master.groupby(["snapshot", "source"]).agg(
    listings=("listing_id", "size"),
    usable=("usable", "sum"),
    suburbs=("suburb_key", "nunique"),
    median_weekly_rent=("weekly_rent", lambda s: s[master.loc[s.index, "usable"]].median()))
overview["% usable"] = (100 * overview["usable"] / overview["listings"]).round(1)
print(f"{len(L.CONTRACT_COLUMNS)} columns: {L.CONTRACT_COLUMNS}")
overview

                 listings  usable  with_feature_list
snapshot source                                     
2025-09  domain     12717   12292              11391
Saved 12,717 listings to data/curated/listings.parquet
62 columns: ['listing_id', 'source', 'snapshot', 'scraped_date', 'suburb', 'postcode', 'suburb_key', 'address', 'lat', 'lon', 'weekly_rent', 'bond', 'bedrooms', 'bathrooms', 'carspaces', 'property_type', 'property_group', 'date_listed', 'days_listed', 'agency', 'photo_count', 'features_text', 'has_air_conditioning', 'has_heating', 'has_dishwasher', 'has_built_in_wardrobes', 'has_balcony_deck', 'has_courtyard_garden', 'has_secure_parking', 'has_ensuite', 'has_study', 'has_pool', 'has_gym', 'has_furnished', 'has_pets_allowed', 'has_floorboards', 'has_internal_laundry', 'has_solar', 'has_security', 'has_outdoor_entertaining', 'has_gas_appliances', 'has_bath', 'has_broadband', 'has_shed', 'has_carpet', 'has_openable_windows', 'has_rumpus_room', 'has_views', 'has_separate_dining',

,,listings,usable,suburbs,median_weekly_rent,% usable
snapshot,source,,,,,
2025-09,domain,12717,12292,654,565.0,96.7


## 2.3 Why listings were set aside

The first table counts each set-aside listing once, by the first Rule A check it failed (`exclusion_reason`). These
rows stay in the table, so an analysis can deliberately relax a rule (a map doesn't need bedrooms). The second table
counts the nine review flags from Part 1 for each year (a listing can have several; flags mark listings for review
and don't decide `usable` on their own).

In [47]:
reasons = (master["exclusion_reason"].fillna("usable")
           .groupby(master["snapshot"]).value_counts().unstack(0).fillna(0).astype(int))
display(reasons.loc[["usable"] + [r for r in L.REASONS if r in reasons.index]])
master.groupby("snapshot")[L.FLAG_COLUMNS].sum().T

snapshot,2025-09
exclusion_reason,
usable,12292
not residential,8
rent missing,278
"rent outside $100-$10,000",53
bad coordinates,4
bedrooms missing,77
implausible room counts,5


snapshot,2025-09
flag_future_listed,0
flag_long_listing,492
flag_invalid_available_date,68
flag_rent_suspicious,53
flag_bedrooms_suspicious,1
flag_bathrooms_suspicious,3
flag_carspaces_suspicious,9
flag_property_features_suspicious,13
flag_bad_coordinates,4


## 2.4 Property types → groups

Every property type as each website writes it, and the group it was placed in. Any type the rules don't recognise is
listed underneath; it should be added to `PROPERTY_GROUPS` in `scripts/listings.py` rather than dropped.

In [48]:
types = L.type_report(master)
display(types)
new_types = L.unknown_types(master)
print("Types not recognised:", new_types.to_dict() if len(new_types) else "none")

,source,property_type,property_group,listings
0,domain,House,House,6589
1,domain,Apartment / Unit / Flat,Apartment/Unit,4516
2,domain,Townhouse,Townhouse,1278
3,domain,Studio,Studio,209
4,domain,New Apartments / Off the Plan,Apartment/Unit,45
5,domain,Villa,House,28
6,domain,Acreage / Semi-Rural,House,13
7,domain,Semi-Detached,House,10
8,domain,New House & Land,House,8
9,domain,Car Space,Not residential,7


Types not recognised: none


## 2.5 The 28 features

For each year: the share of listings where a feature is mentioned (% yes) and where it is unknown (% unknown).
"Unknown" includes listings with no feature list, and, for 2026, the 10 features rent.com.au barely offers.

In [49]:
cover = {}
for snap, g in master.groupby("snapshot"):
    f = g[L.FEATURE_COLUMNS]
    cover[(snap, "% yes")] = (f == 1).mean().mul(100).round(1)
    cover[(snap, "% unknown")] = f.isna().mean().mul(100).round(1)
cover = pd.DataFrame(cover)
cover.index = cover.index.str.replace("has_", "")
cover

2025-09          
                       % yes % unknown
air_conditioning        42.3      10.4
heating                 61.2      10.4
dishwasher              42.9      10.4
built_in_wardrobes      52.6      10.4
balcony_deck            25.1      10.4
courtyard_garden        16.8      10.4
secure_parking          43.1      10.4
ensuite                 14.1      10.4
study                   10.1      10.4
pool                     4.9      10.4
gym                      6.0      10.4
furnished                8.9      10.4
pets_allowed             8.2      10.4
floorboards             15.0      10.4
internal_laundry        32.7      10.4
solar                    2.8      10.4
security                15.7      10.4
outdoor_entertaining     6.0      10.4
gas_appliances          21.9      10.4
bath                    16.8      10.4
broadband                7.3      10.4
shed                     7.3      10.4
carpet                   1.8      10.4
openable_windows         1.5      10.4
rumpus_room              1.4      10.4
views                    2.0      10.4
separate_dining          0.8      10.4
double_glazing           0.8      10.4

Which feature tags were used, per source, and which of the 28 they count towards. The tags that match none of the
28 are listed so nothing slips through unnoticed. "Close to shops / transport / schools" are left out on purpose
(they are the agent's claim; notebook 05 measures real distances), as is "pool/spa count: 0" (it means no pool).

In [50]:
report = L.feature_report(master)
share = report.assign(matched=report["counts_towards"] != "unmatched").groupby(["source", "matched"])["listings"].sum()
print("Share of tag uses matched to one of the 28:",
      (share.xs(True, level="matched") / share.groupby(level="source").sum()).round(3).to_dict())
display(report[report["counts_towards"] == "unmatched"].groupby("source").head(10).reset_index(drop=True))

Share of tag uses matched to one of the 28: {'domain': 0.92}


,source,label,counts_towards,listings
0,domain,close to shops,unmatched,965
1,domain,close to transport,unmatched,947
2,domain,close to schools,unmatched,932
3,domain,pool/spa count: 0,unmatched,253
4,domain,car parking surface,unmatched,98
5,domain,car accom: carport,unmatched,85
6,domain,cable or satellite,unmatched,72
7,domain,energy efficient appliances,unmatched,69
8,domain,indoor spa,unmatched,65
9,domain,water tank,unmatched,65


## 2.6 2025 vs 2026, and links to the other datasets

Median weekly rent of usable listings by property group and by bedrooms for each year, and how many suburbs can be
joined to the SA2 crosswalk (notebook 04) and the DFFH rent-history crosswalk (notebook 03) by `suburb_key`.
SA2 data is later attached by each listing's coordinates, so a suburb missing from the crosswalk still gets an SA2.

In [51]:
use = master[master["usable"]]
display(use.pivot_table(index="property_group", columns="snapshot", values="weekly_rent", aggfunc="median"))
display(use[use["bedrooms"].between(0, 5)]
        .pivot_table(index="bedrooms", columns="snapshot", values="weekly_rent", aggfunc="median"))

sa2_keys = set(pd.read_csv(config.CURATED / "suburb_sa2_crosswalk.csv")["suburb_key"])
dffh_keys = set(pd.read_csv(config.DFFH_CROSSWALK)["suburb_key"].dropna())
by_snapshot = {s: set(g["suburb_key"].dropna()) for s, g in use.groupby("snapshot")}
links = pd.DataFrame({s: {"suburbs": len(k),
                          "in SA2 crosswalk (04)": len(k & sa2_keys),
                          "in DFFH crosswalk (03)": len(k & dffh_keys)} for s, k in by_snapshot.items()}).T
if len(by_snapshot) == 2:
    a, b = by_snapshot.values()
    print(f"Suburbs with usable listings in both years: {len(a & b)}")
links

snapshot,2025-09
property_group,
Apartment/Unit,560.0
House,560.0
Studio,350.0
Townhouse,652.5


snapshot,2025-09
bedrooms,
1.0,460.0
2.0,590.0
3.0,550.0
4.0,600.0
5.0,800.0


,suburbs,in SA2 crosswalk (04),in DFFH crosswalk (03)
2025-09,644,644,204


## 2.7 What the later notebooks use

`data/curated/listings.parquet` holds 62 columns: identity (`listing_id`, `source`, `snapshot`, `scraped_date`),
location (`suburb`, `postcode`, `suburb_key`, `address`, `lat`, `lon`), rent and size (`weekly_rent`, `bond`,
`bedrooms`, `bathrooms`, `carspaces`), type and listing details (`property_type`, `property_group`, `date_listed`,
`days_listed`, `agency`, `photo_count`), the original feature text (`features_text`), the 28 `has_*` features,
`features_missing`, the nine review flags from Part 1 (`flag_future_listed`, `flag_long_listing`,
`flag_invalid_available_date`, `flag_rent_suspicious`, `flag_bedrooms_suspicious`, `flag_bathrooms_suspicious`,
`flag_carspaces_suspicious`, `flag_property_features_suspicious`, `flag_bad_coordinates`), `exclusion_reason` and
`usable`.

Later notebooks read it with `listings.load_listings()` and filter on `usable`. Re-run this notebook whenever new
data arrives (for example after `02_scraping` merges new files); everything downstream then includes it.